libpympb/pympb.cpp

```diff
@@ -1699,22 +1708,41 @@ std::vector<mpb_real> mode_solver::get_output_k() {
   return output_k;
 }
 
+/* get_val returns process-specific output for HAVE_MPI: if the "point" (ix, iy, iz; stride)
+   is on the local process, the value of data at that point is returned; otherwise (i.e.
+   point is not on local process) 0.0 is returned: calls to get_val should therefore be
+   followed by sum-reduction via mpi_allreduce_1(..) in the caller (as in interp_val) */
 mpb_real mode_solver::get_val(int ix, int iy, int iz, int nx, int ny, int nz, int last_dim_size,
                               mpb_real *data, int stride, int conjugate) {
-  // #ifdef HAVE_MPI
-  //      CHECK(0, "get-*-point not yet implemented for MPI!");
-  // #else
-  (void)nx;
-  (void)last_dim_size;
-  (void)conjugate;
-  return data[(((ix * ny) + iy) * nz + iz) * stride];
-  // #endif
+  #ifdef HAVE_MPI
+	  /* due to real-space xy=>yx transposition in MPI configuration, we need to
+	     do a little extra work here; see details e.g. in XYZ_LOOP macro */
+    int local_ny = mdata->local_ny; /* dim of local process over y-indices */
+    int local_y_start = mdata->local_y_start;
+    int local_iy = iy - local_y_start;
+	  mpb_real val = 0; /* reduce local processes over this variable later */
+
+	  /* check if local_iy is in the current process' data block */
+    if (local_iy >= 0 && local_iy < local_ny) { 
+         val = data[(((local_iy * nx) + ix) * nz + iz) * stride]; /* note transposition in x and y indices */
+	  }
+
+  #else /* no MPI */
+    mpb_real val = data[(((ix * ny) + iy) * nz + iz) * stride];
+  #endif
+
+  // TODO: this is in MPB's implementation, but seems to cause issues here - why?
+  // if (conjugate)
+  //   return -val;
+  // else
+  //   return val;
```

Comment:
In MPB's implementation of `get_val`, the `conjugate` keyword works like this. In the previous implementation here though, the `conjugate` keyword is ignored.
When I was testing this, I found that I needed to ignore it here as well for things to turn out right - but I don't understand why there's this difference. Any explanation would be helpful!

Reply:
In MPB's implementation, the `conjugate` keyword is modified here: https://github.com/NanoComp/mpb/blob/74fb432d31905a43b2d2c1d9b88a9f4bbfd0a8b6/mpb/fields.c#L567

That is, a `conjugate == 1` argument is ignored unless (1) MPB is compiled `--with-inv-symmetry` (`! SCALAR_COMPLEX`) and (2) the requested point is in the non-stored "other half" of the data (which is conjugated).     Basically, it corresponds to real-antisymmetric data.    

This isn't relevant in the Python interface because it doesn't support `--with-inv-symmetry` IIRC.

Reply:
It seemed to have an impact here though: if I have the above lines uncommented, `get_bloch_field_point` seemed to return the conjugate of the right answer.

Reply:
That's because you are missing the logic from [this line](https://github.com/NanoComp/mpb/blob/74fb432d31905a43b2d2c1d9b88a9f4bbfd0a8b6/mpb/fields.c#L567).

Reply:
Ahh, that's some bad copy-pasting on my part - thanks for figuring that one out..!
Should I port that bit over, even though the Python interface currently doesn't support `--with-inv-symmetry` (e.g., in case that becomes a desire)?

Reply:
No, I would just omit it, but add a comment to that effect.

Reply:
Similarly for the MPI stuff — since right now, it can't be tested, I'm reluctant to include it.

Reply:
Okay, I've done that in [c743e25](https://github.com/NanoComp/meep/pull/1930/commits/c743e2521581a773c52d76e7f9bb6822b3fa0aef).